### PR TITLE
[processing] Allow configuration of order of outputs created by a model

### LINKED ISSUE #44184: Ability to control the order of processing model output layers
**Feature description.**

There are a lot of use cases where a specific layer order for processing model output layers is wanted, e.g. so that a layer with point features is not hidden by a layer with polygon features. 

Unless I'm missing something, there's currently no way to control the order of processing model output layers, so it could be that you have to rearrange the layers manually after the model is finished.

Here's a [test_model_output.zip](https://github.com/qgis/QGIS/files/6824112/test_model_output.zip) where it looks like the output order is totally random, despite the algorithm `2` is always executed after `1` by setting a dependency.

image

layer_order

Maybe one option to control the layer order would be to make the output order dependent on the order in which the model algorithms are executed. This way it would be possible to control the layer order by setting algorithm dependencies, e.g. for the sample model above layer `2` would be always positioned on top of layer `1`.

**Additional context**

This is a recurring question on SE:

- https://gis.stackexchange.com/questions/157702/processing-modeler-output-order
- https://gis.stackexchange.com/questions/278256/control-rendering-layer-order-of-outputs-from-qgis-3-processing-modeler
- https://gis.stackexchange.com/questions/387284/setting-layer-order-for-output-in-qgis-graphical-modeler
- https://gis.stackexchange.com/questions/377868/output-layers-order-in-qgis-graphic-modeller-algorithm-execution-order-depend

### PR DESCRIPTION
**Pull request by nyalldawson**
This adds a new "Reorder Model Outputs" action to the model designer menu (to accompany the existing "Reorder Model Inputs" action). Selecting this option allows model creators to set a specific order which the outputs from their model must use when loading the results into a project. This gives the model creator a means of ensuring that layers are logically ordered, eg placing a vector layer over a raster layer and a point layer over a polygon layer.

Optionally, the model creator can also set a "Group name" for the outputs. If this is specified then all outputs from the model will be placed into a (newly created if necessary) layer tree group with that name.

Sponsored by the QGIS Germany User Group

Fixes https://github.com/qgis/QGIS/issues/44184

### CODE REVIEW

**Review by pathmapper — COMMENTED** [review]
Very nice to see this feature getting implemented. 

Did some testing with several algorithm outputs/different layer types, also with/without existing group - everything works great 👍

**Review by alexbruy — APPROVED** [review]

File: src/ui/processing/qgsmodeldesignerdialogbase.ui [hunk 1]
```diff
@@ -756,6 +757,11 @@
     <string>New Model…</string>
    </property>
   </action>
+  <action name="mActionReorderOutputs">
+   <property name="text">
+    <string>Reorder Model Outputs...</string>
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/ui/processing/qgsmodeldesignerdialogbase.ui:762]: How about "Reorder Output Layers"? Or "Set Output Layer Order" ?

[the remark this answers is not in the document]
Comment (pathmapper) [reply src/ui/processing/qgsmodeldesignerdialogbase.ui:762]: "Reorder Output Layers" sounds good!

[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/ui/processing/qgsmodeldesignerdialogbase.ui]: Done!

**Comment by dingausmwald** [comment]
Is it possible to use variables in group name? Tried but didn't work. Would be great to enter the group name via modeler input.

**Pull request merged**